### PR TITLE
otelzap: Allow context injection via fields

### PR DESCRIPTION
**Pull request by khushijain21**
Part of https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5191

Pre-work https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5279

### CODE REVIEW

File: bridges/otelzap/core.go [hunk 1]
```diff
@@ -155,17 +158,21 @@ func (o *Core) Write(ent zapcore.Entry, fields []zapcore.Field) error {
...
+		r.AddAttributes(convertField(&o.ctx, fields)...)
 	}
 
-	o.logger.Emit(context.Background(), r)
+	o.logger.Emit(o.ctx, r)
 	return nil
 }
 
-func convertField(fields []zapcore.Field) []log.KeyValue {
+func convertField(ctx *context.Context, fields []zapcore.Field) []log.KeyValue {
```
Comment (pellared) [inline bridges/otelzap/core.go:168]: Can we return the context instead?

```suggestion
func convertField(fields []zapcore.Field) (context.Context, []log.KeyValue) {
```

**Review by pellared — COMMENTED** [review]
We should have at least 2 tests where the context is passed using `.With` or directly as a field.
These could be added as test cases to `TestCore` and with names like `WriteContext` and `WithContext`.

@dmathieu, I see that it will be problematic as the Recorder in https://pkg.go.dev/go.opentelemetry.io/otel/log/logtest does not capture the context. Are you able to modify https://pkg.go.dev/go.opentelemetry.io/otel/log/logtest#ScopeRecords to allow it?

File: bridges/otelzap/core.go [hunk 2]
```diff
@@ -152,31 +154,32 @@ func (o *Core) Write(ent zapcore.Entry, fields []zapcore.Field) error {
...
+		o.ctx, attrbuf = convertField(o.ctx, fields)
+		r.AddAttributes(attrbuf...)
 	}
 
 	o.logger.Emit(o.ctx, r)
 	return nil
 }
 
-func convertField(ctx *context.Context, fields []zapcore.Field) []log.KeyValue {
+func convertField(ctx context.Context, fields []zapcore.Field) (context.Context, []log.KeyValue) {
```
Comment (pellared) [inline bridges/otelzap/core.go:171]: Do we need to pass `ctx context.Context`?

File: bridges/otelzap/core.go [hunk 3]
```diff
@@ -152,31 +154,32 @@ func (o *Core) Write(ent zapcore.Entry, fields []zapcore.Field) error {
...
 	// TODO: Handle attributes passed via With (exceptions: context.Context and zap.Namespace).
-	// TODO: Handle context.Context containing trace context.
 	// TODO: Handle zap.Namespace.
 	// TODO: Handle ent.LoggerName.
 
 	r.AddAttributes(o.attr...)
 	if len(fields) > 0 {
-		r.AddAttributes(convertField(&o.ctx, fields)...)
+		var attrbuf []log.KeyValue
+		o.ctx, attrbuf = convertField(o.ctx, fields)
```
Comment (pellared) [inline bridges/otelzap/core.go:163]: should it not be something like the following?

```suggestion
		ctx, attrbuf := convertField(fields)
		if ctx != nil { 
			o.ctx = ctx
		}
```

File: bridges/otelzap/core.go [hunk 4]
```diff
@@ -116,7 +116,9 @@ func (o *Core) Enabled(level zapcore.Level) bool {
 func (o *Core) With(fields []zapcore.Field) zapcore.Core {
 	cloned := o.clone()
 	if len(fields) > 0 {
-		cloned.attr = append(cloned.attr, convertField(&cloned.ctx, fields)...)
+		var attrbuf []log.KeyValue
+		cloned.ctx, attrbuf = convertField(cloned.ctx, fields)
+		cloned.attr = append(cloned.attr, attrbuf...)
```
Comment (pellared) [inline bridges/otelzap/core.go:121]: should it not be something like the following?


```suggestion
		ctx, attrbuf := convertField(fields)
		if ctx != nil { 
			cloned.ctx = ctx
		}
		cloned.attr = append(cloned.attr, attrbuf...)
```

File: bridges/otelzap/core_test.go [hunk 5]
```diff
@@ -44,6 +44,9 @@ func TestCore(t *testing.T) {
 
 	rec.Reset()
 
+	// TODO WriteContext
+	// TOOD WithContext
```
Comment (pellared) [inline bridges/otelzap/core_test.go:48]:
```suggestion
	// TODO: Add WriteContext test case
	// TOOD: Add WithContext test case
```

File: bridges/otelzap/core.go (hunk 3 above)
Comment (khushijain21) [reply bridges/otelzap/core.go:163]: oh yes you are right, we don't need to pass context

**Review by pellared — APPROVED** [review]

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Pull request merged**